src/ledger/events.rs

```diff
@@ -0,0 +1,95 @@
+use chrono::DateTime;
+use chrono::Utc;
+use ethereum_types::U256;
+use revm::primitives::Address;
+use uuid::Uuid;
+
+use crate::eth::primitives::BlockNumber;
+use crate::eth::primitives::Hash;
+
+/// Represents token transfers (debits and credits) associated with a specific Ethereum account within a single transaction.
+///
+/// The `account_address` field identifies the main account involved in these transfers.
+///
+/// An event will be generated for each account involved in a transaction, meaning if a transaction is not the main account in this event,
+/// in another event it will be and credit and debit parties will be adjusted accordingly.
+pub struct AccountTransfersEvent {
+    /// ID of the event publication.
+    ///
+    /// If the event is republished, a new ID will be generated for the publication.
+    ///
+    /// Format: UUID v7 (xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx)
+    pub publication_id: Uuid,
+
+    /// Datetime of the event publication.
+    ///
+    /// Format: ISO 8601
+    pub publication_datetime: DateTime<Utc>,
+
+    /// Idempotency key of the event payload (TBD: `account_address` + `transaction_hash`).
+    ///
+    /// It is unique for each distinct event, but consistent across retries or republications of the same event payload.
+    ///
+    /// Format: String (TBD)
+    pub idempotency_key: String,
+
+    /// Address of the account that is part of all transfers.
+    ///
+    /// Format: Prefixed account address - 20 bytes - 0x1234567890123456789012345678901234567890
+    pub account_address: Address,
+
+    /// Hash of the Ethereum transaction that originated transfers.
+    ///
+    /// Format: Prefixed transaction hash - 32 bytes - 0x1234567890123456789012345678901234567890123456789012345678901234
+    pub transaction_hash: Hash,
+
+    /// Address of the contract that originated transfers.
+    ///
+    /// Format: Prefixed account address - 20 bytes - 0x1234567890123456789012345678901234567890
+    pub contract_address: Address,
+
+    /// Identifier of the Ethereum function that originated transfers.
+    ///
+    /// Format: Prefixed function signature - 4 bytes - 0x12345678
+    pub function_id: [u8; 4],
+
+    /// Number of the block that originated transfers.
+    ///
+    /// Format: Integer
+    pub block_number: BlockNumber,
+
+    /// Datetime of the block that originated account transfers.
+    ///
+    /// Format: ISO 8601
+    pub block_datetime: DateTime<Utc>,
+
+    /// List of transfers the related `account_address` is part of.
+    pub transfers: Vec<AccountTransfer>,
+}
+
+/// Represents a token transfer between a debit party and a credit party.
+pub struct AccountTransfer {
```

Comment:
I believe this format of debit_party and credit_party works well, but my main concern is that we will inevitably need to infer a some of information from them.
For instance: We will need to compare, for each transfer, both addresses with `AccountTransferEvent.account_address` to determine the direction of the transfer (credit or debit)
Also, the one that is not equal to the account_address will determine the type of transfer.

I believe it would be nice if these two inferences were already done by the producer

Reply:
Good point.
I think I will add a `kind` field inside each transfer with the possible values being `debit` and `credit`?
`debit` means the `account_address` is the `debit_party`, `credit` means it is the `credit_party`.

Reply:
@gabrielpioto do you have an example of how it would be built?

Reply:
@dinhani-cw I prefer the term `direction` over `kind`. In my opinion `kind` is "too generic"
@vicenzi-cw What do you mean? For `direction` we have all information we need to infer in the message. For the type we will need a mapping somewhere of addresses and the event_type. Maybe as a env var / config file

Reply:
Added it as `direction`.

Reply:
sry, you can ignore my comment, @gabrielpioto . I was thinking about single-entry and made some comments that doesnt make sense

Reply:
I think having the fields be named `from` and `to` instead of `credit_party_address` and `debit_party_address` is just as clear and more consistent with the language used in other places, such as in the Transfer event emitted by ERC20 for instance.
I'm not arguing we should be stuck to blockchain lingo, but in this case I find it more readable (I might be biased since we've spent this year focused in developing stratus)

Reply:
@carneiro-cw 

Credit party and debit party are common names used by some financial institutions.
* [Pismo](https://developers.pismo.io/events/docs/integrated-payments-transfer-out-created-1)
* [BTG](https://developer.btgpactual.com/doctr#section/External-Transfer-Status)

Consider the scenario where we want to evolve the API and publish additional information like:
* `debit_party_account_type` / `credit_party_account_type` (normal account or contract account)
* `debit_party_previous_balance` /  `credit_party_previous_balance` (balance before operation)

It would feel more natural than having it as `from_account_type`, `from_previous_balance`, `to_account_type` or `to_previous_balance`.

Reply:
I imagined they were common names, but I think from cloudwalk's POV, since we already implemented the whole transactional structure on top of the blockchain, From and To are even more common.

As for the second point: If the data is flattened I agree. But If it is structured (such as in the BTG example you linked), I think having something like:
```
"from": {
      "account": "000000",
      "accountType": "CC",
      "bank": "208",
      "branch": "1",
      "name": "BTG Pactual",
      "personType": "J",
      "taxId": "00000000000000"
    },
```
Still feels natural (to me). So I stand on my point of preserving naming consistency as much as we can.

Either way I think the naming is appropriate, I just fear that having multiple terms to mean the same thing will cause points of confusion, specially with those less involved in the development of the system's that decide the terms.

Reply:
To me `from_account_type` is more natural than `debit_party_account_type`, even if it's grammatically weird.

But over both, I prefer `sender`, `receiver`, `sender_account_type` and `receiver_account_type`.

Reply:
In this case I concede that credit_party_<suffix> and debit_party_<suffix> reads more naturally, and since it is an industry standard it does makes sense to use it.
(I still like `from` and `to` better though :p mainly for naming consistency across different systems)